README.md


A Dart repository for testing io mocking in Dart, supports various Medium articles, 
see [this](https://medium.com/@stevehamblett/mocking-in-dart-using-iooverrides-a-gem-of-a-class-d2838b80a1a8) medium article for an example.



CHANGELOG.md
## 0.2.0

- Dart 3

## 0.1.0

- Initial version.


test/mockio_test.dart
import 'package:mockio/mockio.dart';
import 'package:test/test.dart';

void main() {
  group('A group of tests', () {
    final awesome = Awesome();

    setUp(() {
      // Additional setup goes here.
    });

    test('First Test', () {
      expect(awesome.isAwesome, isTrue);
    });
  });
}


test/socket_io_test.dart
import 'dart:io';
import 'dart:async';
import 'dart:typed_data';

import 'package:test/test.dart';
import 'package:mocktail/mocktail.dart';
import 'package:event_bus/event_bus.dart';

class MockSocket extends Mock implements Socket {
  final eventBus = EventBus();
  final mockBytes = <int>[];
  MockSocket() {
    eventBus.on().listen((event) {
      // Print the runtime type. Such a set up could be used for logging.
      print('Hello from MockSocket - ${event.runtimeType}');
    });
  }
  static Future<MockSocket> connect(host, int port,
      {sourceAddress, int sourcePort = 0, Duration? timeout}) {
    final completer = Completer<MockSocket>();
    final extsocket = MockSocket();
    completer.complete(extsocket);
    return completer.future;
  }
}

final mySocket = MockSocket();

void main() {
  group('Group 1 - Simple', () {
    setUp(() {
      // Additional setup goes here.
    });

    test('Getter', () {
      when(() => mySocket.port).thenReturn(500);
      when(() => mySocket.remotePort).thenReturn(600);
      expect(mySocket.port, 500);
      expect(mySocket.remotePort, 600);
    });

    test('Setter', () {
      when(() => mySocket.setOption(SocketOption.tcpNoDelay, any()))
          .thenReturn(true);
      expect(mySocket.setOption(SocketOption.tcpNoDelay, true), isTrue);
    });
  });

  group('Group 2 - Dynamic', () {
    setUp(() {
      // Additional setup goes here.
    });

    test('Add', () {
      when(() => mySocket.add(any())).thenAnswer((invocation) {
        mySocket.mockBytes.addAll(invocation.positionalArguments[0]);
      });
      mySocket.add([1, 2, 3]);
      expect(mySocket.mockBytes, [1, 2, 3]);
    });

    test('Take', () async {
      when(() => mySocket.take(any())).thenAnswer((invocation) {
        final length = invocation.positionalArguments[0];
        final taken = mySocket.mockBytes.take(length).toList();
        mySocket.mockBytes.removeRange(0, length);
        final utaken = Uint8List.fromList(taken);
        final out = StreamController<Uint8List>();
        out.add(utaken);
        out.done;
        out.close();
        return out.stream;
      });
      final taken = mySocket.take(2);
      expect(await taken.single, [1, 2]);
      expect(mySocket.mockBytes, [3]);
    });
  });

  group('Group 3 - Connect', () {
    test('Simple', () {
      const myPort = 1000;
      const myHost = 'mine';
      IOOverrides.runZoned(() async {
        var mySocket = await Socket.connect(myHost, myPort);
        when(() => mySocket.port).thenReturn(myPort);
        expect(mySocket is MockSocket, isTrue);
        mySocket = mySocket as MockSocket;
        final eb = mySocket.eventBus;
        eb.fire('Hello');
        expect(mySocket.port, myPort);
      },
          socketConnect: (dynamic host, int port,
              {dynamic sourceAddress,
                int sourcePort = 0,
                Duration? timeout}) =>
              MockSocket.connect(host, port,
                  sourceAddress: sourceAddress,
                  sourcePort: sourcePort,
                  timeout: timeout));
    });
    test('Listen', () {
      const myPort = 1000;
      const myHost = 'mine';
      IOOverrides.runZoned(() async {
        var mySocket = await Socket.connect(myHost, myPort);
        when(() => mySocket.port).thenReturn(myPort);
        when(() => mySocket.add(any())).thenAnswer((invocation) {
          (mySocket as MockSocket)
              .mockBytes
              .addAll(invocation.positionalArguments[0]);
        });
        expect(mySocket is MockSocket, isTrue);
        mySocket = mySocket as MockSocket;
        final eb = mySocket.eventBus;
        eb.fire('Hello');
        expect(mySocket.port, myPort);
        mySocket.add([1, 2, 3]);
        mySocket.listen((event) {});
      },
          socketConnect: (dynamic host, int port,
              {dynamic sourceAddress,
                int sourcePort = 0,
                Duration? timeout}) =>
              MockSocket.connect(host, port,
                  sourceAddress: sourceAddress,
                  sourcePort: sourcePort,
                  timeout: timeout));
    });
  });
}


test/mockio_socket_io_test.dart
/*
 * Package : mockio
 * Author : S. Hamblett <steve.hamblett@linux.com>
 * Date   : 08/03/2023
 * Copyright :  S.Hamblett
 */

import 'dart:io';

import 'package:test/test.dart';
import 'package:mocktail/mocktail.dart';
import 'package:mockio/mockio.dart';

final mySocket = MockSocket();

void main() {
  group('Group 1 - Simple', () {
    setUp(() {
      // Additional setup goes here.
    });

    test('Getter', () {
      expect(mySocket.port, 0);
      expect(mySocket.host, '');
    });

    test('Setter', () {
      when(() => mySocket.setOption(SocketOption.tcpNoDelay, any()))
          .thenReturn(true);
      expect(mySocket.setOption(SocketOption.tcpNoDelay, true), isTrue);
    });
  });

  group('Group 2 - Dynamic', () {
    setUp(() {
      // Additional setup goes here.
    });

    test('Add', () {
      mySocket.add([1, 2, 3]);
      expect(mySocket.mockBytes, [1, 2, 3]);
    });
  });

  group('Group 3 - Connect', () {
    test('Simple', () {
      const myPort = 1000;
      const myHost = 'mine';
      IOOverrides.runZoned(() async {
        var mySocket = await Socket.connect(myHost, myPort);
        expect(mySocket is MockSocket, isTrue);
        mySocket = mySocket as MockSocket;
        expect(mySocket.port, myPort);
        expect(mySocket.host, myHost);
      },
          socketConnect: (dynamic host, int port,
                  {dynamic sourceAddress,
                  int sourcePort = 0,
                  Duration? timeout}) =>
              MockSocket.connect(host, port,
                  sourceAddress: sourceAddress,
                  sourcePort: sourcePort,
                  timeout: timeout));
    });
    test('Listen', () {
      const myPort = 1000;
      const myHost = 'mine';
      IOOverrides.runZoned(() async {
        var mySocket = await Socket.connect(myHost, myPort);
        expect(mySocket is MockSocket, isTrue);
        mySocket = mySocket as MockSocket;
        expect(mySocket.port, myPort);
        expect(mySocket.host, myHost);
        mySocket.add([1, 2, 3]);
        mySocket.listen((data) {
          expect(data, [1, 2, 3]);
        });
      },
          socketConnect: (dynamic host, int port,
                  {dynamic sourceAddress,
                  int sourcePort = 0,
                  Duration? timeout}) =>
              MockSocket.connect(host, port,
                  sourceAddress: sourceAddress,
                  sourcePort: sourcePort,
                  timeout: timeout));
    });
  });
}


lib/src/mockio_mock_socket.dart
/*
 * Package : mockio
 * Author : S. Hamblett <steve.hamblett@linux.com>
 * Date   : 08/03/2023
 * Copyright :  S.Hamblett
 */

import 'dart:async';
import 'dart:io';
import 'dart:typed_data';

import 'package:mocktail/mocktail.dart';
import 'package:mqtt_client/mqtt_client.dart';
import 'package:typed_data/typed_data.dart';

//
/// The mock socket class
///
class MockSocket extends Mock implements Socket {
  final mockBytes = <int>[];
  final mockBytesUint = Uint8List(500);

  dynamic onDataFunc;
  dynamic onDoneFunc;
  dynamic onErrorFunc;

  MockSocket();

  static late MockSocket instance;

  @override
  int port = 0;

  String host = '';

  late StreamSubscription<Uint8List> outgoing =
      Stream<Uint8List>.empty().listen((event) {});

  static Future<MockSocket> connect(host, int port,
      {sourceAddress, int sourcePort = 0, Duration? timeout}) {
    final completer = Completer<MockSocket>();
    final extSocket = MockSocket();
    extSocket.port = port;
    extSocket.host = host;
    instance = extSocket;
    completer.complete(extSocket);
    return completer.future;
  }

  @override
  void add(List<int> data) {
    mockBytes.addAll(data);
  }

  @override
  StreamSubscription<Uint8List> listen(void Function(Uint8List event)? onData,
      {Function? onError, void Function()? onDone, bool? cancelOnError}) {
    onDataFunc = onData;
    onDoneFunc = onDone;
    onErrorFunc = onError;

    final out = Uint8List.fromList(mockBytes);
    onData!(out);
    return outgoing;
  }

  @override
  void destroy();

  @override
  Future close() {
    final completer = Completer<Future>();
    return completer.future;
  }

  void onDone() => onDoneFunc();
}

///
/// Mock socket scenario class
///
class MqttScenario1 extends MockSocket {
  static Future<MqttScenario1> connect(host, int port,
      {sourceAddress, int sourcePort = 0, Duration? timeout}) {
    final completer = Completer<MqttScenario1>();
    final extSocket = MqttScenario1();
    extSocket.port = port;
    extSocket.host = host;
    MockSocket.instance = extSocket;
    completer.complete(extSocket);
    return completer.future;
  }

  @override
  void add(List<int> data) {
    mockBytes.addAll(data);
    final ack = MqttConnectAckMessage()
        .withReturnCode(MqttConnectReturnCode.connectionAccepted);
    final buff = Uint8Buffer();
    final ms = MqttByteBuffer(buff);
    ack.writeTo(ms);
    ms.seek(0);
    final out = Uint8List.fromList(ms.buffer!.toList());
    onDataFunc(out);
  }

  @override
  StreamSubscription<Uint8List> listen(void Function(Uint8List event)? onData,
      {Function? onError, void Function()? onDone, bool? cancelOnError}) {
    onDataFunc = onData;
    onDoneFunc = onDone;
    onErrorFunc = onError;
    return outgoing;
  }
}
